subgraphs/cbridge/src/mappings.ts

```diff
@@ -0,0 +1,1187 @@
+    protocol._cumulativeRewardsClaimed = event.params.reward;
+    protocol.save();
+    return;
+  } else if (
+    event.block.timestamp <
+    protocol._lastRewardTimestamp!.plus(
+      BigInt.fromI32(SECONDS_PER_DAY * averageRewardOverDays)
+    )
+  ) {
+    protocol._cumulativeRewardsClaimed =
+      protocol._cumulativeRewardsClaimed!.plus(event.params.reward);
+    protocol.save();
+    return;
+  }
+
+  // allocate rewards to pool proportional to tvl
+  const rToken = sdk.Tokens.getOrCreateToken(event.params.token);
+  const poolIDs = protocol._liquidityPoolIDs!;
+  // first iteration summing tvl
+  let sumTVLUSD = BIGDECIMAL_ZERO;
+  for (let i = 0; i < poolIDs.length; i++) {
+    const pool = sdk.Pools.loadPool(poolIDs[i]).pool;
+    sumTVLUSD = sumTVLUSD.plus(pool.totalValueLockedUSD);
+  }
+
+  for (let i = 0; i < poolIDs.length; i++) {
+    const pool = sdk.Pools.loadPool(poolIDs[i]);
+    const poolRewardAmount = bigDecimalToBigInt(
+      pool.pool.totalValueLockedUSD
+        .div(sumTVLUSD)
+        .times(protocol._cumulativeRewardsClaimed!.toBigDecimal())
+        .div(BigDecimal.fromString(averageRewardOverDays.toString()))
+    );
+    pool.setRewardEmissions(RewardTokenType.DEPOSIT, rToken, poolRewardAmount);
+  }
+
+  protocol._lastRewardTimestamp = event.block.timestamp;
+  protocol._cumulativeRewardsClaimed = BIGINT_ZERO;
+  protocol.save();
+}
+
+// Pegged Token Bridge V2
+export function handlePTBv2Burn(event: PTBv2Burn): void {
+  const poolId = event.address.concat(event.params.token);
+  const networkConstants = getNetworkSpecificConstant(event.params.toChainId);
+  const dstPoolId = networkConstants.getPoolAddress(
+    PoolName.OriginalTokenVaultV2
+  );
+
+  _handleTransferOut(
+    event.params.token, //srcToken
+    event.params.account,
+    event.params.toAccount,
+    event.params.amount,
+    event.params.toChainId,
+    poolId,
+    dstPoolId,
+    BridgePoolType.BURN_MINT,
+    CrosschainTokenType.WRAPPED,
+    event,
+    event.params.burnId
+  );
+}
+
+export function handleMessage2(event: Message2): void {
+  const customEvent = _createCustomEvent(event);
+  if (!customEvent) {
+    log.error("[handleMessage2]customeEvent cannot be null", []);
+    return;
+  }
+  _handleMessageOut(
+    event.params.dstChainId,
+    event.params.sender,
+    Address.fromBytes(event.params.receiver), //this may truncate addresses for non-EVM chain
+    event.params.message,
+    event.params.fee,
+    customEvent
+  );
+}
+
+export function handleMessage(event: Message): void {
+  const customEvent = _createCustomEvent(event);
+  if (!customEvent) {
+    log.error("[handleMessage]customeEvent cannot be null", []);
+    return;
+  }
+  _handleMessageOut(
+    event.params.dstChainId,
+    event.params.sender,
+    event.params.receiver,
+    event.params.message,
+    event.params.fee,
+    customEvent
+  );
+}
+
+export function handleMessageWithTransfer(event: MessageWithTransfer): void {
```

Comment:
Does this do anything else? ie, do we need to use `_handleTransferOut()`?

Reply:
good question. My reading of their contract is that users will need to do a separate `sendTokenTransfer` call when sendMessageWithTransfer, which will emit a transfer out event (which event depending on the tyep of bridge used): https://github.com/celer-network/sgn-v2-contracts/blob/9ce8ffe13389415a53e2c38838da1b99689d40f0/contracts/message/libraries/MessageSenderLib.sol#L66-L98.

I added a comment for this.